### ISSUE TITLE
[one-optimize] Optimize Mul/Reshape/Div pattern

### ISSUE THREAD

**Issue by BalyshevArtem** [Samsung/ONE#11956]
## What 

Let's optimize Reshape/Mean/Mul/Reshape/Div pattern, where Mul(x_mul, y_mul), y_mul - is const scalar value and Div(x_div, y_div), x_div - is const scalar value:

![original](https://github.com/Samsung/ONE/assets/43214667/79d12d8e-1b99-4ab2-ad29-534438e0b6f2)

as Mean/Div where Div(x_new, y_new), x_new const scalar value equal to (x_div / x_mul):

![Screenshot from 2023-11-10 13-44-09](https://github.com/Samsung/ONE/assets/43214667/c19f284c-4bdf-4ef5-9675-e6bf3633b7d6)

## Why 

Our target model has this pattern

**Comment by jinevening** [comment]
@BalyshevArtem Could you give a background of this work?

The draft (#11995) would work only for a very specific pattern. If this task is urgent, it would be ok to proceed with it. However, we typically make a more generally-applicable pass. For example, it would be possible to forward Reshape Op across Mean/Mul/Div. Then, two Reshape Ops will be fused, so the final Ops will become Mean->Div->Reshape.

**Comment by BalyshevArtem** [comment]
> @BalyshevArtem Could you give a background of this work?
> 
> The draft (#11995) would work only for a very specific pattern. If this task is urgent, it would be ok to proceed with it. However, we typically make a more generally-applicable pass. For example, it would be possible to forward Reshape Op across Mean/Mul/Div. Then, two Reshape Ops will be fused, so the final Ops will become Mean->Div->Reshape.

No, nothing urgent, just found such pattern in some target models. So, better co create 3 passes: 1)forward Reshape op across Mean/Mul/Div, 2)  Then fusing two reshape ops, 3) And fusing Mul/Div as Div, right?

**Comment by jinevening** [comment]
We already have 2). 1) and 3) would be enough.

**Comment by BalyshevArtem** [comment]
> We already have 2). 1) and 3) would be enough.

Introduce new #12124 draft. To optimize this pass I added CircleDiv, CircleMul, CircleMean to forwardReshape op pass (https://github.com/Samsung/ONE/pull/12124#discussion_r1406314803, https://github.com/Samsung/ONE/pull/12124#discussion_r1406315209, https://github.com/Samsung/ONE/pull/12124#discussion_r1406318707).
Then introduce new pass to fuse Div and Mul as Div. So using --fuse_mul_div --remove_unnecessary_reshape --remove_redundant_reshape --forward_reshape_to_unaryop, optimize this pattern.

![Screenshot from 2023-11-27 18-26-51](https://github.com/Samsung/ONE/assets/43214667/83fb3be3-8777-47ea-a47c-2bab7d869364)


![Screenshot from 2023-11-27 18-27-16](https://github.com/Samsung/ONE/assets/43214667/8c12eed0-3f9f-4763-8b9d-ae40e57ff8b9)

**Comment by seanshpark** [comment]
Q) @jinevening , for our NPU, which is better `Mul` or `Div`? or don't care?

**Comment by jinevening** [comment]
@seanshpark Our NPU supports `Mul` better than `Div` in terms of supported patterns and optimization passes.

@BalyshevArtem Thanks for the work. Could you update `--fuse_mul_div` to generate `Mul` instead of `Div`? I found that Tensorflow also favors Mul over Div. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/tensorflow/transforms/canonicalize.td#L162

**Comment by seanshpark** [comment]
@jinevening  thanks for the search!

**Comment by BalyshevArtem** [comment]
> Could you update `--fuse_mul_div` to generate `Mul` instead of `Div`? I found that Tensorflow also favors Mul over Div.

I see, but the main problem that `Div` in this pattern smth like this `div_const / input_from_mul`,  where `input_from_mul = mul_const * input_from_mean`. So, if we will fuse `div_const` and `mul_const` into one Mul we will need anyway new Div = `1 / input_from_mul'. So, that is why I used here fusing to div

**Comment by seanshpark** [comment]
From the graph, `out = Div(output_of_mean, DIV_Const)`
can't we get `out = Mul(output_of_mean, MUL_Const)` where `Mul_Const = 1.0 / Div_Const` ?

**Comment by BalyshevArtem** [comment]
> From the `Mean - Div - Reshape` graph, `out = Div(output_of_mean, DIV_Const)`
> can't we get `out = Mul(output_of_mean, MUL_Const)` where `Mul_Const = 1.0 / Div_Const` ?

From such graph, yes, we can. We can replace DIV(some_input, DIV_const) to MUL(som_input, 1/DIV_const) - I can write such pass. 
But in pattern above in DIV op const goes in first place: DIV(DIV_const, some_input).

**Comment by seanshpark** [comment]
> DIV(DIV_const, some_input).

OK, got it. thanks for the explanation. we have to go as-is.

**Comment by jinevening** [comment]
Ah, I misunderstood the pattern. Thanks for clarification.

**Comment by seanshpark** [comment]
@BalyshevArtem ,

You have added
- `fuse_mul_with_div` to `circle2circle` and
- `fuse_mul_div` to `one-cmds`.
Why did this happen? 
1/ please find the reason why this happened had share so that we can prevent this again
2/ please fix this as I assume CI failure (#11777) maybe cause of this

CC @jinevening

**Comment by BalyshevArtem** [comment]
posted #12314

@seanshpark, thank you!

**Comment by seanshpark** [comment]
Why did this happen?
1/ please find the reason why this happened had share so that we can prevent this again
2/ please fix this as I assume CI failure (https://github.com/Samsung/ONE/issues/11777) maybe cause of this

**Comment by BalyshevArtem** [comment]
> Why did this happen? 1/ please find the reason why this happened ~had~ and share so that we can prevent this again

I am not sure, I think I just copy it from draft, and in the draft I had comment (https://github.com/Samsung/ONE/pull/12124#discussion_r1408590506) to change it from `fuse_mul_div` to `fuse_mul_with_div` and I forgot to change it in `one-cmds` part in the draft. I am sorry

**Comment by seanshpark** [comment]
seems done
